src/app/clusters/camera-av-settings-user-level-management-server/camera-av-settings-user-level-management-server.cpp

```diff
@@ -0,0 +1,1079 @@
+                                         Optional(static_cast<uint8_t>(newZoom)));
+
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    // Set the local values of pan, tilt, and zoom
+    SetPan(Optional<int16_t>(newPan));
+    SetTilt(Optional<int16_t>(newTilt));
+    SetZoom(Optional<uint8_t>(static_cast<uint8_t>(newZoom)));
+
+    MarkDirty(Attributes::MPTZPosition::Id);
+
+    ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+}
+
+void CameraAvSettingsUserLevelMgmtServer::HandleMPTZMoveToPreset(HandlerContext & ctx,
+                                                                 const Commands::MPTZMoveToPreset::DecodableType & commandData)
+{
+    Status status  = Status::Success;
+    uint8_t preset = commandData.presetID;
+
+    // Verify the provided presetID is within spec limits
+    //
+    if (preset > mMaxPresets - 1)
+    {
+        ChipLogError(Zcl, "Preset provided is out of range");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    // Do we have any presets?
+    //
+    if (mMptzPresetHelpers.empty())
+    {
+        ChipLogError(Zcl, "No stored presets, MoveToPreset not possible");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    // We have presets, check that the received ID is a valid preset ID
+    //
+    auto it = std::find_if(mMptzPresetHelpers.begin(), mMptzPresetHelpers.end(),
+                           [preset](const MPTZPresetHelper & mptzph) { return mptzph.GetPresetID() == preset; });
+
+    if (it == mMptzPresetHelpers.end())
+    {
+        ChipLogError(Zcl, "No matching presets, MoveToPreset not possible");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::ConstraintError);
+        return;
+    }
+
+    // This is effectively a manipulation of the current PTZ settings, ensure that the device is in a state wherein a PTZ change is
+    // possible
+    //
+    if (!mDelegate->CanChangeMPTZ())
+    {
+        ChipLogDetail(Zcl, "Device not able to process move to MPTZ preset");
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, Status::Busy);
+        return;
+    }
+
+    auto presetValues = it->GetMptzPosition();
+
+    // Inform the delegate that the device is requested to move to PTZ values given by the selected preset id
+    // Call the delegate to allow the device to handle the physical changes, on success set the MPTZ values based on the preset
+    status = mDelegate->MPTZMoveToPreset(preset, presetValues.pan, presetValues.tilt, presetValues.zoom);
+
+    if (status != Status::Success)
+    {
+        ctx.mCommandHandler.AddStatus(ctx.mRequestPath, status);
+        return;
+    }
+
+    mMptzPosition = presetValues;
+    MarkDirty(Attributes::MPTZPosition::Id);
```

Comment:
Should check that something actually changed before marking dirty?

Reply:
Reworked this